app/src/main/java/de/westnordost/streetcomplete/quests/findNodesAtCrossingsOf.kt

```
@@ -0,0 +1,145 @@
+
+        /* So, surrounding the shared node X, in the simple case, we have
+         *
+         * 1. position A, B neighbouring the shared node position which are part of a barrier way
+         * 2. position P, Q neighbouring the shared node position which are part of the moving way
+         *
+         * The way crosses the barrier if P is on one side of the polyline spanned by A,X,B and
+         * Q is on the other side.
+         *
+         * How can a moving way that has a shared node with a barrier not cross the latter?
+         * Multiple separate paths can reach barrier in one shared node - all from one side
+         * Imagine the road at https://www.openstreetmap.org/node/258003112 would continue to
+         * the south here, then, still none of those footways would actually cross the street
+         *  - only if it continued straight to the north, for example.
+         *
+         * The example brings us to the less simple case: What if several barriers share
+         * a node at a crossing-candidate position?
+         * Also, what if there are more than one footways involved?
+         *
+         * We look for if there is ANY crossing, so all polylines involved are checked:
+         * For all barriers going through point X, it is checked if not all passing ways that
+         * go through X are on the same side of the barrier-polyline.
+         * */
+        val nodePos = mapData.getNode(nodeId)?.position
+        return@retainAll nodePos != null &&
+            neighbouringWayPositions.anyCrossesAnyOf(neighbouringBarrierPositions, nodePos)
+    }
+
+    val resultMap = waysByNodeId.mapValues {
+        WaysCrossing(
+            mapData.getNode(it.key)!!,
+            mutableListOf(),
+            it.value
+        )
+    }
+    barriersByNodeId.forEach { (nodeId, barrierWays) ->
+        if (resultMap.contains(nodeId)) {
+            resultMap[nodeId]!!.barrierWays += barrierWays
+        }
+    }
+    return resultMap.values;
```

```suggestion
    return waysByNodeId.map { (nodeId, ways) ->
        WaysCrossing(
            mapData.getNode(nodeId)!!,
            barriersByNodeId.getValue(nodeId),
            ways
        )
    }
```

Shouldn't this do the same and be shorter and easier to read?

---

yes, and it - as far as I can see - should return the same. Still, I would test once more.

---

Wouldn't `barriersByNodeId.getValue(nodeId),` return null if the id is not present, instead of an empty list? Unless it has a default value for the getter.
So I think to keep the same semantics, you need `barriersByNodeId.getValue(nodeId) ?: listOf()`.
Not sure if it matters in practice as I didn't look outside this snippet.

---

At this point we filter only intersections that are defined - among others - by having connection with both barrier and moving way.

---

Okay. In that case, maybe add a `!!`, so we will get crash reports from a NPE instead of surprising behavior, in case the assumption is wrong?